src/uwtools/cli.py

```diff
@@ -0,0 +1,593 @@
+        required=True,
+        type=str,
+    )
+
+
+def add_arg_values_file(group: Group, required: bool = False) -> None:
+    group.add_argument(
+        "--values-file",
+        help="Path to file providing override or interpolation values",
+        metavar="PATH",
+        required=required,
+        type=str,
+    )
+
+
+def add_arg_values_format(group: Group, choices: List[str], required: bool = False) -> None:
+    group.add_argument(
+        "--values-format",
+        choices=choices,
+        help="Values format",
+        required=required,
+        type=str,
+    )
+
+
+def add_arg_values_needed(group: Group) -> None:
+    group.add_argument(
+        "--values-needed",
+        action="store_true",
+        help="Print report of values needed to render template",
+    )
+
+
+def add_arg_verbose(group: Group) -> None:
+    group.add_argument(
+        "--verbose",
+        "-v",
+        action="store_true",
+        help="Print all logging messages",
+    )
+
+
+# pylint: enable=missing-function-docstring
+
+
+# Support
+
+
+def abort(msg: str) -> None:
+    """
+    Exit with an informative message and error status.
+
+    :param msg: The message to print.
+    """
+    print(msg, file=sys.stderr)
+    sys.exit(1)
+
+
+def add_subparser(subparsers: Subparsers, name: str, helpmsg: str) -> Parser:
+    """
+    Add a new subparser, with standard help formatting, to the given parser.
+
+    :param subparsers: The subparsers to add the new subparser to.
+    :param name: The name of the subparser.
+    :param helpmsg: The help message for the subparser.
+    :return: The new subparser.
+    """
+    return subparsers.add_parser(
+        name, add_help=False, help=helpmsg, formatter_class=formatter, description=helpmsg
+    )
+
+
+def add_subparsers(parser: Parser, dest: str) -> Subparsers:
+    """
+    Add subparsers to a parser.
+
+    :parm parser: The parser to add subparsers to.
+    :return: The new subparsers object.
+    """
+    return parser.add_subparsers(
+        dest=dest, metavar="MODE", required=True, title="Positional arguments"
+    )
+
+
+def basic_setup(parser: Parser) -> Group:
+    """
+    Create optional-arguments group and add help switch.
+
+    :param parser: The parser to add the optional group to.
+    """
+    optional = parser.add_argument_group("Optional arguments")
+    optional.add_argument("-h", "--help", action="help", help="Show help and exit")
+    return optional
+
+
+def check_args(args: Namespace) -> Namespace:
+    """
+    Validate basic argument correctness.
+
+    :param args: The parsed command-line arguments to check.
+    :return: The checked command-line arguments.
+    :raises: SystemExit if any checks failed.
+    """
+    a = vars(args)
+    if a.get("quiet") and a.get("verbose"):
+        abort("Specify at most one of --quiet, --verbose")
+    if a.get("values_file") and not a.get("values_format"):
+        abort("Specify --values-format with --values-file")
```

Comment:
I have a follow-on PR that adds that flexibility back, making this optional -- well, optional as long as a file is specified; if we're reading from stdin -- or, conversely, writing to stdout -- then the format is still mandatory, because you can't deduce anything from a non-existent filename. But that check is done after the arguments are parsed (in the follow-on PR). I'll open that PR as soon as this one merges. It's much simpler.